### PR TITLE
[Flutter GPU] Texture binding, index binding, attachments, depth state.

### PR DESCRIPTION
**Pull request by bdero**
Now rendering textured 3D models!

* Combined depth+stencil attachment.
* Allow for multiple color attachments.
* Add blend mode configuration.
* Fix uniform ordering for vertex shaders.
* Texture binding and sampling options.
* Index buffer binding.
* Depth configuration.

https://github.com/flutter/engine/assets/919017/56790431-07aa-4362-b6d3-f5460287cf2b

```cpp
  void paint(Canvas canvas, Size size) {
    /// Allocate a new renderable texture.
    final gpu.Texture? renderTexture = gpu.gpuContext.createTexture(
        gpu.StorageMode.devicePrivate, 300, 300,
        enableRenderTargetUsage: true,
        enableShaderReadUsage: true,
        coordinateSystem: gpu.TextureCoordinateSystem.renderToTexture);
    if (renderTexture == null) {
      return;
    }

    final gpu.Texture? depthTexture = gpu.gpuContext.createTexture(
        gpu.StorageMode.deviceTransient, 300, 300,
        format: gpu.gpuContext.defaultDepthStencilFormat,
        enableRenderTargetUsage: true,
        coordinateSystem: gpu.TextureCoordinateSystem.renderToTexture);
    if (depthTexture == null) {
      return;
    }

    /// Create the command buffer. This will be used to submit all encoded
    /// commands at the end.
    final commandBuffer = gpu.gpuContext.createCommandBuffer();

    /// Define a render target. This is just a collection of attachments that a
    /// RenderPass will write to.
    final renderTarget = gpu.RenderTarget.singleColor(
      colorAttachment: gpu.ColorAttachment(texture: renderTexture),
      depthStencilAttachment: gpu.DepthStencilAttachment(
          texture: depthTexture, depthClearValue: 1.0),
    );

    /// Add a render pass encoder to the command buffer so that we can start
    /// encoding commands.
    final encoder = commandBuffer.createRenderPass(renderTarget);

    /// Load a shader bundle asset.
    final library = gpu.ShaderLibrary.fromAsset('TestLibrary')!;

    /// Create a RenderPipeline using shaders from the asset.
    final vertex = library['TextureVertex']!;
    final fragment = library['TextureFragment']!;
    final pipeline = gpu.gpuContext.createRenderPipeline(vertex, fragment);

    encoder.bindPipeline(pipeline);

    encoder.setDepthWriteEnable(true);
    encoder.setDepthCompareOperation(gpu.CompareFunction.less);

    /// (Optional) Configure blending for the first color attachment.
    encoder.setColorBlendEnable(true);
    encoder.setColorBlendEquation(gpu.ColorBlendEquation(
        colorBlendOperation: gpu.BlendOperation.add,
        sourceColorBlendFactor: gpu.BlendFactor.one,
        destinationColorBlendFactor: gpu.BlendFactor.oneMinusSourceAlpha,
        alphaBlendOperation: gpu.BlendOperation.add,
        sourceAlphaBlendFactor: gpu.BlendFactor.one,
        destinationAlphaBlendFactor: gpu.BlendFactor.oneMinusSourceAlpha));

    /// Append quick geometry and uniforms to a host buffer that will be
    /// automatically uploaded to the GPU later on.
    final transients = gpu.HostBuffer();
    final vertices = transients.emplace(float32(<double>[
      -1, -1, -1, /* */ 0, 0, /* */ 1, 0, 0, 1, //
      1, -1, -1, /*  */ 1, 0, /* */ 0, 1, 0, 1, //
      1, 1, -1, /*   */ 1, 1, /* */ 0, 0, 1, 1, //
      -1, 1, -1, /*  */ 0, 1, /* */ 0, 0, 0, 1, //
      -1, -1, 1, /*  */ 0, 0, /* */ 0, 1, 1, 1, //
      1, -1, 1, /*   */ 1, 0, /* */ 1, 0, 1, 1, //
      1, 1, 1, /*    */ 1, 1, /* */ 1, 1, 0, 1, //
      -1, 1, 1, /*   */ 0, 1, /* */ 1, 1, 1, 1, //
    ]));
    final indices = transients.emplace(uint16(<int>[
      0, 1, 3, 3, 1, 2, //
      1, 5, 2, 2, 5, 6, //
      5, 4, 6, 6, 4, 7, //
      4, 0, 7, 7, 0, 3, //
      3, 2, 7, 7, 2, 6, //
      4, 5, 0, 0, 5, 1, //
    ]));
    final mvp = transients.emplace(float32Mat(Matrix4(
          0.5, 0, 0, 0, //
          0, 0.5, 0, 0, //
          0, 0, 0.2, 0, //
          0, 0, 0.5, 1, //
        ) *
        Matrix4.rotationX(time) *
        Matrix4.rotationY(time * seedX) *
        Matrix4.rotationZ(time * seedY)));

    /// Bind the vertex and index buffer.
    encoder.bindVertexBuffer(vertices, 8);
    encoder.bindIndexBuffer(indices, gpu.IndexType.int16, 36);

    final mvpSlot = pipeline.vertexShader.getUniformSlot('mvp')!;
    encoder.bindUniform(mvpSlot, mvp);

    final sampledTexture = gpu.gpuContext.createTexture(
        gpu.StorageMode.hostVisible, 5, 5,
        enableShaderReadUsage: true);
    sampledTexture!.overwrite(uint32(<int>[
      0xFFFFFFFF, 0x00000000, 0xFFFFFFFF, 0x00000000, 0xFFFFFFFF, //
      0x00000000, 0xFFFFFFFF, 0x00000000, 0xFFFFFFFF, 0x00000000, //
      0xFFFFFFFF, 0x00000000, 0xFFFFFFFF, 0x00000000, 0xFFFFFFFF, //
      0x00000000, 0xFFFFFFFF, 0x00000000, 0xFFFFFFFF, 0x00000000, //
      0xFFFFFFFF, 0x00000000, 0xFFFFFFFF, 0x00000000, 0xFFFFFFFF, //
    ]));

    final texSlot = pipeline.fragmentShader.getUniformSlot('tex')!;
    encoder.bindTexture(texSlot, sampledTexture);

    /// And finally, we append a draw call.
    encoder.draw();

    /// Submit all of the previously encoded passes. Passes are encoded in the
    /// same order they were created in.
    commandBuffer.submit();

    /// Wrap the Flutter GPU texture as a ui.Image and draw it like normal!
    final image = renderTexture.asImage();

    canvas.drawImage(image, Offset(-renderTexture.width / 2, 0), Paint());
  }
```

TextureVertex
```glsl
uniform mat4 mvp;

in vec3 position;
in vec2 texture_coords;
in vec4 color;
out vec2 v_texture_coords;
out vec4 v_color;

void main() {
  v_texture_coords = texture_coords;
  v_color = color;
  gl_Position = mvp * vec4(position, 1.0);
}
```

TextureFragment
```glsl
uniform sampler2D tex;

in vec2 v_texture_coords;
in vec4 v_color;
out vec4 frag_color;

void main() {
  frag_color = v_color * texture(tex, v_texture_coords);
}
```

### CODE REVIEW

**Review by jonahwilliams — APPROVED** [review]
Some nits on the Dart API

Overall I dislike the pattern of validating in the constructor of an Object via thrown exception. Not sure what the best alternative would be.

File: lib/gpu/fixtures.h [hunk 1]
```diff
@@ -40,3 +40,70 @@ extern unsigned char kFlutterGPUUnlitVertIPLR[];
...
+  };
+  static constexpr std::array<const impeller::ShaderStageBufferLayout*, 1>
+      kInterleavedBufferLayout = {&kInterleavedLayout};
+};
+
+constexpr unsigned int kFlutterGPUTextureVertIPLRLength = 920;
+extern unsigned char kFlutterGPUTextureVertIPLR[];
+
+constexpr unsigned int kFlutterGPUTextureFragIPLRLength = 800;
+extern unsigned char kFlutterGPUTextureFragIPLR[];
```
[the remark this answers is not in the document]
Comment (bdero) [reply lib/gpu/fixtures.h]: Done.

File: lib/gpu/lib/src/render_pass.dart [hunk 2]
```diff
@@ -7,58 +7,129 @@
...
+  MinMagFilter minFilter;
+  MinMagFilter magFilter;
+  MipFilter mipFilter;
+  SamplerAddressMode widthAddressMode;
+  SamplerAddressMode heightAddressMode;
+}
+
+base class RenderTarget {
+  RenderTarget(
+      {List<ColorAttachment>? colorAttachments, this.depthStencilAttachment}) {
```
[the remark this answers is not in the document]
Comment (bdero) [reply lib/gpu/lib/src/render_pass.dart]: Done. Nice one!

File: lib/gpu/lib/src/render_pass.dart [hunk 3]
```diff
@@ -7,58 +7,129 @@
...
+      {ColorAttachment? colorAttachment, this.depthStencilAttachment}) {
+    if (colorAttachment != null) {
+      colorAttachments = <ColorAttachment>[colorAttachment];
+      return;
+    }
+    colorAttachments = <ColorAttachment>[];
+  }
+
+  late List<ColorAttachment> colorAttachments;
+  DepthStencilAttachment? depthStencilAttachment;
```
[the remark this answers is not in the document]
Comment (bdero) [reply lib/gpu/lib/src/render_pass.dart]: Done. I converted `RenderTarget.singleColor` into a non-const factory. I _think_ that's necessary, but let me know if I'm missing some const magic.

File: lib/gpu/lib/src/render_pass.dart [hunk 4]
```diff
@@ -7,58 +7,127 @@
...
+      {ColorAttachment? colorAttachment,
+      DepthStencilAttachment? depthStencilAttachment}) {
+    List<ColorAttachment> colors = [];
+    if (colorAttachment != null) {
+      colors.add(colorAttachment);
+    }
+    return new RenderTarget(
+        colorAttachments: colors,
+        depthStencilAttachment: depthStencilAttachment);
+  }
```
Comment (jonahwilliams) [inline lib/gpu/lib/src/render_pass.dart:98]: With Dart's limited const expression support you won't be able to make this const, but it doesn't need to be a factory either.

It seems odd that the colorAttachment argument to this object is nullable?
```suggestion
    return RenderTarget(
        colorAttachments: [
        
        ],
        depthStencilAttachment: depthStencilAttachment);
  }
```

File: lib/gpu/lib/src/render_pass.dart [hunk 5]
```diff
@@ -7,58 +7,127 @@
...
+      {this.colorAttachments = const <ColorAttachment>[],
+      this.depthStencilAttachment});
+
+  factory RenderTarget.singleColor(
+      {ColorAttachment? colorAttachment,
+      DepthStencilAttachment? depthStencilAttachment}) {
+    List<ColorAttachment> colors = [];
+    if (colorAttachment != null) {
+      colors.add(colorAttachment);
+    }
```
Comment (jonahwilliams) [inline lib/gpu/lib/src/render_pass.dart:94]:
```suggestion
```

Comment (bdero) [reply lib/gpu/lib/src/render_pass.dart]: Done.

**Comment by zanderso** [comment]
> Overall I dislike the pattern of validating in the constructor of an Object via thrown exception. Not sure what the best alternative would be.

A pattern that I like better (but don't necessarily love) is to have a factory constructor return an "invalid" instance that stashes a description of the error state somewhere when something goes wrong during construction. Sort of like this: https://github.com/flutter/engine/blob/main/tools/pkg/engine_build_configs/lib/src/build_config.dart#L92. Then the caller can decide whether to throw an exception after checking `obj.valid` or whatever.

**Comment by jonahwilliams** [comment]
I think it would be more reasonable to throw expceptions than to use an isValid bit. My problem with the valid bit is that it is easy to ignore/forget, and then the developer only gets the error when the invalid object is passed to a different method instead of at the original invalidation.

I would also investigate how WebGPU does validation for render target/attachment descriptions.

**Comment by bdero** [comment]
This is part of the Flutter GPU MVP umbrella bug: https://github.com/flutter/flutter/issues/130921

**Pull request merged**